### PR TITLE
tools: Add initial dev-doctor

### PR DESCRIPTION
**Pull request by twpayne**
This PR adds a command that diagnoses your development setup for common problems which you can execute by running:

```
make dev-doctor
```

The goal is to:
* Assist contributors (or potential contributors) by helping them find problems in their development setup.
* Assist maintainers by reducing the burden of diagnosing problems in contributor's development setups.

The current code is a starting point and includes the following checks:
* Supported operating system and architecture.
* Required binaries are installed and at least the minimum required version.
* The user's groups are configured correctly.
* NFS and `iptables` are configured correctly.

Output on my Linux box looks like this:

```
RESULT    CHECK            MESSAGE
ok        os-arch          linux/amd64
ok        go               found /usr/bin/go, version 1.15.2
ok        clang            found /usr/bin/clang, version 10.0.0
ok        ginkgo           found /home/twp/bin/ginkgo, version 1.14.1
ok        golangci-lint    found /home/twp/bin/golangci-lint, version 1.31.0
ok        docker           found /usr/bin/docker
warning   docker-compose   exec: "docker-compose": executable file not found in $PATH
ok        vagrant          found /usr/bin/vagrant, version 2.2.6
ok        virtualbox       found /usr/bin/virtualbox
ok        pip3             found /usr/bin/pip3, version 20.0.2
ok        docker-group     twp in docker group
ok        /etc/nfs.conf    nfsd.tcp is "y"
ok        iptables-rule    INPUT -p tcp -s 192.168.34.0/24 --dport 111 -j ACCEPT
ok        iptables-rule    INPUT -p tcp -s 192.168.34.0/24 --dport 2049 -j ACCEPT
ok        iptables-rule    INPUT -p tcp -s 192.168.34.0/24 --dport 20048 -j ACCEPT

See https://docs.cilium.io/en/latest/contributing/development/dev_setup/
```

Whereas on my not-so-well-supported MacBook Pro it looks like:

```
RESULT    CHECK            MESSAGE
warning   os-arch          darwin/amd64
ok        go               found /usr/local/bin/go, version 1.15.3
ok        clang            found /usr/bin/clang, version 12.0.0
ok        ginkgo           found /Users/tompayne/bin/ginkgo, version 1.14.2
ok        golangci-lint    found /Users/tompayne/bin/golangci-lint, version 1.31.1-0.20201018110117-e4346a8bd6b3
ok        docker           found /usr/local/bin/docker
ok        docker-compose   found /usr/local/bin/docker-compose
ok        vagrant          found /usr/local/bin/vagrant, version 2.2.10
ok        virtualbox       found /usr/local/bin/virtualbox
ok        pip3             found /usr/local/bin/pip3, version 20.2.3
skipped   docker-group     docker group only used on linux
error     /etc/nfs.conf    nfsd.tcp is not set, want "y"
skipped   iptables-rule    iptables only used on linux
skipped   iptables-rule    iptables only used on linux
skipped   iptables-rule    iptables only used on linux

See https://docs.cilium.io/en/latest/contributing/development/dev_setup/
exit status 1
```

The current checks are a reasonable start but far from complete. That said, it should be relatively straightforward to add extra checks as needed.

### CODE REVIEW

**Comment by sayboras** [comment]
Hope you don't mind if I chime in. Please find below input from my system.

Please ignore golangci-lint check as I usually built from source. Just curious about /etc/nfs.conf and iptables-rule are for which use case/actions. FYI, I can bring up k8s cluster with NFS via something like `make -j && NFS=1 K8S=1 NWORKERS=1 NO_BUILD=1 ./contrib/vagrant/start.sh` locally.

```shell script
$ make dev-doctor
go run ./tools/dev-doctor
RESULT   CHECK           MESSAGE
ok       os-arch         linux/amd64
ok       go              found /snap/go/current/bin/go, version 1.15.3
ok       clang           found /usr/bin/clang, version 10.0.0
ok       ginkgo          found /home/tammach/go/bin/ginkgo, version 1.12.0
failed   golangci-lint   found /home/tammach/go/bin/golangci-lint, could not parse version from golangci-lint has version (devel) built from (unknown, mod sum: "") on (unknown)

iptables: Bad rule (does a matching rule exist in that chain?).
iptables: Bad rule (does a matching rule exist in that chain?).
iptables: Bad rule (does a matching rule exist in that chain?).
ok       docker           found /usr/bin/docker
ok       docker-compose   found /home/tammach/.local/bin/docker-compose
ok       vagrant          found /usr/bin/vagrant, version 2.2.9
ok       virtualbox       found /usr/bin/virtualbox
ok       pip3             found /usr/bin/pip3, version 20.0.2
ok       docker-group     tammach in docker group
error    /etc/nfs.conf    /etc/nfs.conf does not exist
failed   iptables-rule    exit status 1
failed   iptables-rule    exit status 1
failed   iptables-rule    exit status 1

See https://docs.cilium.io/en/latest/contributing/development/dev_setup/
exit status 1
make: *** [Makefile:603: dev-doctor] Error 1
```

File: tools/dev-doctor/main.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+func main() {
+	checks := []check{
+		osArchCheck{},
+		&binaryCheck{
+			name:          "go",
+			versionArgs:   []string{"version"},
+			versionRegexp: regexp.MustCompile(`go version go(\d+\.\d+\.\d+)`),
+			minVersion:    &semver.Version{Major: 1, Minor: 15, Patch: 0},
```
Comment (tklauser) [inline tools/dev-doctor/main.go]: Something for a follow-up: Would be nice to somehow be able to bump this automatically upon new Go major releases using the `make update-golang` target.

File: tools/dev-doctor/binarycheck.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
+package main
```
Comment (aanm) [inline tools/dev-doctor/binarycheck.go:15]: all go files in this directory miss the copyright header

**Review by aanm — CHANGES_REQUESTED** [review]
It would be nice to leave an INFO message to inform the developer why this script asked me for my sudo password.

Also, it assumes the developer is running `iptables` which is not always the case. For example, Fedora >= 32 is installed with nftables, the command to check for the rules is the following:

```
$ sudo nft list ruleset | grep 192.168
                ip saddr 192.168.34.0/24 tcp dport 111 ct state { new, untracked } accept
                ip saddr 192.168.34.0/24 tcp dport 2049 ct state { new, untracked } accept
                ip saddr 192.168.34.0/24 tcp dport 20048 ct state { new, untracked } accept
```
for which the developer can "fix" the issue by running `sudo nft -f contrib/vagrant/nftables.rules`

File: tools/dev-doctor/iptablescheck.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+		cmd = exec.Command("iptables", append([]string{"-C"}, []string(c.rule)...)...)
+		cmd.Stdin = nil
+		cmd.Stdout = nil
+		cmd.Stderr = nil
+
+	} else {
+		cmd = exec.Command("sudo", append([]string{"iptables", "-C"}, []string(c.rule)...)...)
+		cmd.Stdin = os.Stdin
+		cmd.Stdout = nil
+		cmd.Stderr = os.Stderr
```
Comment (pchaigno) [inline tools/dev-doctor/iptablescheck.go]: Why is stderr printed when running as an unprivileged users but not when running as root? I'm getting the following warnings which I would expect to be silenced since we already print a message for missing iptables rules.
```
iptables: Bad rule (does a matching rule exist in that chain?).
iptables: Bad rule (does a matching rule exist in that chain?).
iptables: Bad rule (does a matching rule exist in that chain?).
```

File: tools/dev-doctor/etcnfsconfcheck.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+// An etcdNFSConfCheck checks /etc/nfs.conf.
+type etcNFSConfCheck struct{}
+
+func (etcNFSConfCheck) id() string {
+	return "/etc/nfs.conf"
+}
+
+func (etcNFSConfCheck) run() (checkResult, string) {
+	data, err := ioutil.ReadFile("/etc/nfs.conf")
```
Comment (pchaigno) [inline tools/dev-doctor/etcnfsconfcheck.go:33]: What is this file and why is it needed? I don't have it, but my development and test VMs work fine with NFS.

File: tools/dev-doctor/main.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	checkFailed:  "failed",
+}
+
+func main() {
+	checks := []check{
+		osArchCheck{},
+		&binaryCheck{
+			name:          "go",
+			versionArgs:   []string{"version"},
+			versionRegexp: regexp.MustCompile(`go version go(\d+\.\d+\.\d+)`),
```
Comment (pchaigno) [inline tools/dev-doctor/main.go]: I don't have a patch version number to my output. I think we can safely ignore that last digit anyway.
```
$ go version
go version go1.14 linux/amd64
```

File: tools/dev-doctor/main.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+		&binaryCheck{
+			name:          "ginkgo",
+			versionArgs:   []string{"version"},
+			versionRegexp: regexp.MustCompile(`Ginkgo Version (\d+\.\d+\.\d+)`),
+			minVersion:    &semver.Version{Major: 1, Minor: 4, Patch: 0},
+		},
+		// FIXME add gomega check?
+		&binaryCheck{
+			name:          "golangci-lint",
+			versionArgs:   []string{"--version"},
```
Comment (pchaigno) [inline tools/dev-doctor/main.go]: Weird, this flag is unknown for my `golangci-lint` and `version` gives no information. I can double check how I installed it if that helps.
That being said, do we really need a version check or did we include the version more as a safeguard? /cc @sayboras 
```
$ golangci-lint --version
Error: unknown flag: --version
[...]
$ golangci-lint version
golangci-lint has version master built from ? on
```

File: tools/dev-doctor/main.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+		etcNFSConfCheck{},
+		&iptablesRuleCheck{
+			rule: []string{"INPUT", "-p", "tcp", "-s", "192.168.34.0/24", "--dport", "111", "-j", "ACCEPT"},
+		},
+		&iptablesRuleCheck{
+			rule: []string{"INPUT", "-p", "tcp", "-s", "192.168.34.0/24", "--dport", "2049", "-j", "ACCEPT"},
+		},
+		&iptablesRuleCheck{
+			rule: []string{"INPUT", "-p", "tcp", "-s", "192.168.34.0/24", "--dport", "20048", "-j", "ACCEPT"},
+		},
```
Comment (pchaigno) [inline tools/dev-doctor/main.go]: The iptables/ufw/nft rules are only required if a firewall is enabled. That may be a "corporate requirement" for some users, but we probably don't want to make all other users think this is required.

**Review by pchaigno — CHANGES_REQUESTED** [review]
That's a great idea! :tada: 

A couple comments below, from running on my own setup.

File: tools/dev-doctor/main.go (hunk 7 above)
Comment (tklauser) [inline tools/dev-doctor/main.go]: On my system (Ubuntu 20.04), NFS works fine without these rules in place.

File: tools/dev-doctor/etcnfsconfcheck.go [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+	switch {
+	case os.IsNotExist(err):
+		return checkError, "/etc/nfs.conf does not exist"
```
Comment (tklauser) [inline tools/dev-doctor/etcnfsconfcheck.go:36]: NFS seems to work fine on my setup without this file in place 🤔

**Review by tklauser — CHANGES_REQUESTED** [review]
Great idea! I've added some comments inline. Maybe we could relax the checks a bit for the initial version and then incrementally refine them.

File: tools/dev-doctor/main.go (hunk 6 above)
Comment (sayboras) [reply tools/dev-doctor/main.go]: With the current configuration in .golangci.yaml, it's more as a safeguard :)

I just found out yesterday that `--version` flag only available in release (i.e. building from source will not have it)

File: tools/dev-doctor/iptablescheck.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"runtime"
+	"strings"
+)
+
+// An iptablesRuleCheck checks that the give iptables rule is present.
```
Comment (qmonnet) [inline tools/dev-doctor/iptablescheck.go]:
```suggestion
// An iptablesRuleCheck checks that the given iptables rule is present.
```

File: tools/dev-doctor/main.go (hunk 7 above)
Comment (qmonnet) [reply tools/dev-doctor/main.go]: One way to proceed could be to only check for these rules if the default policy for `INPUT` is not `ACCEPT`, maybe?

File: Makefile [hunk 10]
```diff
@@ -599,5 +599,8 @@ update-images-go-version:
 	$(QUIET) $(MAKE) -C images update-golang-image
 	@echo "Updated go version in image Dockerfiles to $(GO_VERSION)"
 
-.PHONY: force generate-api generate-health-api install build-context-update clean-build clean clean-container veryclean
+dev-doctor:
+	go run ./tools/dev-doctor
```
Comment (qmonnet) [inline Makefile]:
```suggestion
	$(QUIET)$(GO) run ./tools/dev-doctor
```

File: Documentation/contributing/development/dev_setup.rst [hunk 11]
```diff
@@ -9,6 +9,16 @@
 Development Setup
 =================
 
+Verifying your development setup
```
Comment (qmonnet) [inline Documentation/contributing/development/dev_setup.rst]: Nit: Most other titles at this level in the document seem to use title case?

**Review by qmonnet — CHANGES_REQUESTED** [review]
Awesome tool!

A few very minor nits, and I share others' considerations on NFS and iptables rules, but looks great overall, and I'm sure it will be super helpful to new developers. Thanks for this.

[Sorry for the split review, I probably clicked the wrong button at some point.]

File: tools/dev-doctor/binarycheck.go (hunk 2 above)
Comment (twpayne) [reply tools/dev-doctor/binarycheck.go:15]: Fixed, thanks.

File: tools/dev-doctor/iptablescheck.go (hunk 3 above)
Comment (twpayne) [reply tools/dev-doctor/iptablescheck.go]: This is because I thought I needed to pass through stderr so that `sudo` could prompt for a password. Fixed now.

File: tools/dev-doctor/etcnfsconfcheck.go (hunk 4 above)
Comment (twpayne) [reply tools/dev-doctor/etcnfsconfcheck.go:33]: This file configures NFS. When you run `contrib/vagrant/start.sh` a message to ensure that nfsd is configured to use TCP if you have a firewall enabled is printed. That said, I've made the NFS check optional.

File: tools/dev-doctor/main.go (hunk 5 above)
Comment (twpayne) [reply tools/dev-doctor/main.go]: Fixed, thanks.

File: tools/dev-doctor/main.go (hunk 6 above)
Comment (twpayne) [reply tools/dev-doctor/main.go]: OK, I've switched to using `golangci-lint version` instead.

File: tools/dev-doctor/main.go (hunk 7 above)
Comment (twpayne) [reply tools/dev-doctor/main.go]: OK, in the short time I've made the NFS firewall checks disabled by default and opt-in.

File: tools/dev-doctor/main.go (hunk 1 above)
Comment (twpayne) [reply tools/dev-doctor/main.go]: Nice idea, thanks. I've added a `update-golang-dev-doctor` target which is depended on by `update-golang`.

File: tools/dev-doctor/main.go (hunk 7 above)
Comment (twpayne) [reply tools/dev-doctor/main.go]: Ack. I've made the NFS checks disabled by default and opt-in.

File: tools/dev-doctor/etcnfsconfcheck.go (hunk 8 above)
Comment (twpayne) [reply tools/dev-doctor/etcnfsconfcheck.go:36]: Done. NFS checks are now opt-in.

File: tools/dev-doctor/iptablescheck.go (hunk 9 above)
Comment (twpayne) [reply tools/dev-doctor/iptablescheck.go]: Thanks, fixed.

File: Makefile (hunk 10 above)
Comment (twpayne) [reply Makefile]: Thanks, done.

File: Documentation/contributing/development/dev_setup.rst (hunk 11 above)
Comment (twpayne) [reply Documentation/contributing/development/dev_setup.rst]: Thanks, done.

File: tools/dev-doctor/check.go [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+// A checkResult is the result of a check.
+type checkResult int
```
Comment (christarazi) [inline tools/dev-doctor/check.go:18]: Nit: I find "check" to be slightly redundant. This variable and others following can probably drop "check" from them, and it'll still be clear enough.

File: tools/dev-doctor/binarycheck.go [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+	"fmt"
+	"os/exec"
+	"regexp"
+
+	"github.com/blang/semver"
+)
+
+// A binaryCheck checks that a binary called name is installed and optionally at
+// least version minVersion.
+type binaryCheck struct {
```
Comment (christarazi) [inline tools/dev-doctor/binarycheck.go:32]: (If you decide to adopt the style nit below, then this would also make sense to just call it "binary". It feels like just calling each "check" by what it stands for, seems enough.)

File: tools/dev-doctor/check.go [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+	checkOK      checkResult = 0  // The check completed and did not find any problems.
+	checkWarning checkResult = 1  // The check completed and found something that might indicate a problem.
+	checkError   checkResult = 2  // The check completed and found a definite problem.
+	checkFailed  checkResult = 3  // The check could not be completed.
+)
+
+// A check is an individual check.
+type check interface {
+	Name() string               // Name returns the check's name.
+	Run() (checkResult, string) // Run runs the check.
```
Comment (christarazi) [inline tools/dev-doctor/check.go:31]: Nit: generally, the last return value is an `error`. Here I notice the `string` represents the error string anyway, which is then printed when the tool finishes all its checks. While this works, it's not a "typical" Go. Alternatively, this code can return `error` (and it won't need to retrieve & return the string representation on each instance), and just simply format the final output once.

File: tools/dev-doctor/rootcmd.go [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+			minVersion:    &semver.Version{Major: 10, Minor: 0, Patch: 0},
+		},
+		// FIXME add llvm check?
+		// FIXME add libelf-devel check?
+		&binaryCheck{
+			name:          "ginkgo",
+			versionArgs:   []string{"version"},
+			versionRegexp: regexp.MustCompile(`Ginkgo Version (\d+\.\d+\S*)`),
+			minVersion:    &semver.Version{Major: 1, Minor: 4, Patch: 0},
+			hint:          `Run "go get -u go get -u github.com/onsi/ginkgo/ginkgo".`,
```
Comment (christarazi) [inline tools/dev-doctor/rootcmd.go]: This seems like a typo :)

**Review by christarazi — CHANGES_REQUESTED** [review]
This is a fantastic idea, thanks for your time and effort on this!

Mostly nits below as other reviewers have reviewed more in-depth.

**Comment by aditighag** [comment]
Very cool! I ran the command on the dev VM, and I don't understand the flagged `iptables` related output. There is no mention of `iptables` in the dev guide?

Here is the output - 

```
 make dev-doctor
go run ./tools/dev-doctor
iptables: Bad rule (does a matching rule exist in that chain?).
iptables: Bad rule (does a matching rule exist in that chain?).
iptables: Bad rule (does a matching rule exist in that chain?).
RESULT    CHECK            MESSAGE
ok        os-arch          linux/amd64
error     go               found /usr/local/bin/go, version 1.14.3, need 1.15.0
ok        clang            found /usr/bin/clang, version 11.0.0
ok        ginkgo           exec: "ginkgo": executable file not found in $PATH
ok        golangci-lint    found /home/vagrant/go/bin/golangci-lint, version 1.27.0
ok        docker           found /usr/bin/docker
ok        docker-compose   found /usr/local/bin/docker-compose
ok        vagrant          exec: "vagrant": executable file not found in $PATH
warning   virtualbox       exec: "virtualbox": executable file not found in $PATH
ok        pip3             found /usr/bin/pip3, version 9.0.1
ok        docker-group     vagrant in docker group
error     /etc/nfs.conf    /etc/nfs.conf does not exist
failed    iptables-rule    exit status 1
failed    iptables-rule    exit status 1
failed    iptables-rule    exit status 1

See https://docs.cilium.io/en/latest/contributing/development/dev_setup/
exit status 1
Makefile:603: recipe for target 'dev-doctor' failed
make: *** [dev-doctor] Error 1
```

**Comment by twpayne** [comment]
Thank you for the feedback everybody. I've updated the PR with the following changes:
* Improved version checks.
* NFS firewall checks are now disabled by default and opt-in (you need to pass the `--nfs-firewall` option to enable them).
* Added  a `uname` check.
* Added hints on how to fix, where relevant.
* Used `github.com/spf13/cobra`.
* Added copyright headers to all source files.
* Use `$(GO)` to run Go in Makefile.
* Added automatic update of minimum Go version to `update-golang` make target.

The output now looks something like this (note that the `sudo` prompt is there because the `--nfs-firewall` option is specified, and this is documented in the help, normally you don't need `sudo`):

```
$ go run ./tools/dev-doctor --nfs-firewall
[sudo] password for twp: 
RESULT    CHECK            MESSAGE
ok        os/arch          linux/amd64
ok        uname            Linux thinkpad 5.4.0-52-generic #57-Ubuntu SMP Thu Oct 15 10:57:00 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
ok        go               found /usr/bin/go, version 1.15.2
ok        clang            found /usr/bin/clang, version 10.0.0-4ubuntu1
warning   ginkgo           ginkgo not found in $PATH
ok        golangci-lint    found /home/twp/bin/golangci-lint, version 1.31.0
ok        docker           found /usr/bin/docker
warning   docker-compose   docker-compose not found in $PATH
ok        vagrant          found /usr/bin/vagrant, version 2.2.6
ok        virtualbox       found /usr/bin/virtualbox
ok        pip3             found /usr/bin/pip3, version 20.0.2
ok        docker-group     user twp in docker group
error     /etc/nfs.conf    /etc/nfs.conf does not exist
error     iptables-rule    rule INPUT -p tcp -s 192.168.34.0/24 --dport 111 -j ACCEPT not found
ok        iptables-rule    found rule INPUT -p tcp -s 192.168.34.0/24 --dport 2049 -j ACCEPT
ok        iptables-rule    found rule INPUT -p tcp -s 192.168.34.0/24 --dport 20048 -j ACCEPT

CHECK           HINT
ginkgo          Run "go get -u github.com/onsi/ginkgo/ginkgo".
/etc/nfs.conf   Ensure that /etc/nfs.conf includes "[nfsd]\ntcp=y\n".
iptables-rule   Run "sudo iptables -A INPUT -p tcp -s 192.168.34.0/24 --dport 111 -j ACCEPT".

See https://docs.cilium.io/en/latest/contributing/development/dev_setup/.
exit status 1
```

**Review by qmonnet — APPROVED** [review]
Everything looks in order from my side :+1:

File: tools/dev-doctor/check.go (hunk 14 above)
Comment (twpayne) [reply tools/dev-doctor/check.go:31]: The execution flow in this program with respect to errors is a little different to normal. Normally, when you encounter an error you abandon what you're doing and return to the caller. When running all these checks, you want to keep going and move on to the next check, even if an individual check fails.

There's also a difference here that the last value returned is not an error, it is a message to be shown to the user.

In summary, a check does not return an error in the usual Go sense of an error. It always returns a result and a message, and in the worst case that result can be "failed" and the message explains why.

Hope this makes sense.

File: tools/dev-doctor/binarycheck.go (hunk 13 above)
Comment (twpayne) [reply tools/dev-doctor/binarycheck.go:32]: See comment on nit below.

**Comment by qmonnet** [comment]
Oh I ran the command one more time, and I now see the following (this didn't happen on your first version).
```
failed   clang            Prerelease is empty
```

My clang version:

```
$ clang --version
Ubuntu clang version 11.0.0-++20200715081714+92f7bd3a399-1~exp1~20200715062327.261
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

File: tools/dev-doctor/check.go (hunk 12 above)
Comment (twpayne) [reply tools/dev-doctor/check.go:18]: Personally, I prefer the inclusion of `check` in the types as it links the concepts together. `binaryCheck`, `unameCheck`, and `checkResult` are clearly related, whereas `binary`, `uname`, and `result` are not.

File: tools/dev-doctor/rootcmd.go (hunk 15 above)
Comment (twpayne) [reply tools/dev-doctor/rootcmd.go]: Thanks, fixed.

**Review by christarazi — APPROVED** [review]
LGTM

**Comment by twpayne** [comment]
> I ran the command on the dev VM, and I don't understand the flagged `iptables` related output. There is no mention of `iptables` in the dev guide?

@aditighag Instructions about allowing NFS through a firewall are printed by `contrib/vagrant/start.sh`. They're not mentioned in the dev guide.



> Oh I ran the command one more time, and I now see the following (this didn't happen on your first version).

@qmonnet Thanks for spotting this. The version number is parsed with `github.com/blang/semver` as that's the semantic versioning library that we're already using. This library gets confused by the version `11.0.0-++20200715081714+92f7bd3a399-1~exp1~20200715062327.261`. I've "fixed" this by only passing the initial major.minor.patch part of the version to `semver` and using `semver.ParseTolerant`.

**Comment by qmonnet** [comment]
> I've "fixed" this by only passing...

Great! `make dev-doctor` is now all green for me :tada:

**Review by jrajahalme — APPROVED** [review]

File: tools/dev-doctor/etcnfsconfcheck.go (hunk 4 above)
Comment (pchaigno) [reply tools/dev-doctor/etcnfsconfcheck.go:33]: Where do you see that message? Here's what I have:
```
$ K8S=1 NO_BUILD=1 NETNEXT=1 ./contrib/vagrant/start.sh 
# NFS enabled. don't forget to enable these ports on your host
# before starting the VMs in order to have nfs working
# iptables -I INPUT -p tcp -s 192.168.34.0/24 --dport 111 -j ACCEPT
# iptables -I INPUT -p tcp -s 192.168.34.0/24 --dport 2049 -j ACCEPT
# iptables -I INPUT -p tcp -s 192.168.34.0/24 --dport 20048 -j ACCEPT
# To use kubectl on the host, you need to add the following route:
# ip route add 192.168.33.11 via 192.168.34.11
Bringing machine 'k8s1+' up with 'virtualbox' provider...
==> k8s1+: Importing base box 'cilium/ubuntu-next'...
==> k8s1+: Matching MAC address for NAT networking...
==> k8s1+: Setting the name of the VM: cilium_k8s1_1603820305549_67317
==> k8s1+: Clearing any previously set network interfaces...
==> k8s1+: Preparing network interfaces based on configuration...
    k8s1+: Adapter 1: nat
    k8s1+: Adapter 2: intnet
    k8s1+: Adapter 3: hostonly
==> k8s1+: Forwarding ports...
    k8s1+: 6443 (guest) => 7443 (host) (adapter 1)
    k8s1+: 9081 (guest) => 9081 (host) (adapter 1)
    k8s1+: 2345 (guest) => 2345 (host) (adapter 1)
    k8s1+: 22 (guest) => 2222 (host) (adapter 1)
==> k8s1+: Running 'pre-boot' VM customizations...
==> k8s1+: Booting VM...
==> k8s1+: Waiting for machine to boot. This may take a few minutes...
    k8s1+: SSH address: 127.0.0.1:2222
    k8s1+: SSH username: vagrant
    k8s1+: SSH auth method: private key
    k8s1+: 
    k8s1+: Vagrant insecure key detected. Vagrant will automatically replace
    k8s1+: this with a newly generated keypair for better security.
    k8s1+: 
    k8s1+: Inserting generated public key within guest...
    k8s1+: Removing insecure key from the guest if it's present...
    k8s1+: Key inserted! Disconnecting and reconnecting using new SSH key...
==> k8s1+: Machine booted and ready!
==> k8s1+: Checking for guest additions in VM...
==> k8s1+: Setting hostname...
==> k8s1+: Configuring and enabling network interfaces...
==> k8s1+: Exporting NFS shared folders...
==> k8s1+: Preparing to edit /etc/exports. Administrator privileges will be required...
==> k8s1+: Mounting NFS shared folders...
==> k8s1+: Running provisioner: bootstrap (shell)...
```

**Comment by twpayne** [comment]
@aanm For the Fedora/nftables support, I propose to add that in a follow-up PR, if needed. NFS firewall tests are now disabled by default.

**Review by sayboras — APPROVED** [review]
Awesome work. LGTM for my setup now :)

```
$ make dev-doctor
go run ./tools/dev-doctor
RESULT   CHECK            MESSAGE
ok       os/arch          linux/amd64
ok       uname            Linux ubuntu20 5.4.0-52-generic #57-Ubuntu SMP Thu Oct 15 10:57:00 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
ok       go               found /snap/bin/go, version 1.15.3
ok       clang            found /usr/bin/clang, version 10.0.0
ok       ginkgo           found /home/tammach/go/bin/ginkgo, version 1.12.0
ok       golangci-lint    found /home/tammach/go/bin/golangci-lint, version 1.32.0
ok       docker           found /usr/bin/docker
ok       docker-compose   found /home/tammach/.local/bin/docker-compose
ok       vagrant          found /usr/bin/vagrant, version 2.2.9
ok       virtualbox       found /usr/bin/virtualbox
ok       pip3             found /usr/bin/pip3, version 20.0.2
ok       docker-group     user tammach in docker group
```

**Review by tklauser — APPROVED** [review]
Very nice, LGTM. Thanks for adding the `update-golang-dev-doctor` target!

File: tools/dev-doctor/rootcmd.go [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+		},
+		&binaryCheck{
+			name:          "vagrant",
+			ifNotFound:    checkWarning,
+			versionArgs:   []string{"--version"},
+			versionRegexp: regexp.MustCompile(`Vagrant (\d+\.\d+\S*)`),
+			minVersion:    &semver.Version{Major: 2, Minor: 0, Patch: 0},
+		},
+		&binaryCheck{
+			name:       "virtualbox",
```
Comment (vadorovsky) [inline tools/dev-doctor/rootcmd.go]: Can we make that check case insensitive? Currently I'm getting a warning despite the fact, that I have VirtualBox installed as `/usr/bin/VirtualBox`:

```
❯ make dev-doctor
go run ./tools/dev-doctor
RESULT    CHECK            MESSAGE
ok        os/arch          linux/amd64
ok        uname            Linux new-host-6 5.9.0-rc5-1-default+ #2 SMP Mon Oct 26 23:41:40 CET 2020 x86_64 x86_64 x86_64 GNU/Linux
ok        go               found /usr/bin/go, version 1.15.3
ok        clang            found /home/vadorovsky/repos/llvm-project/llvm/build/bin/clang, version 12.0.0
ok        ginkgo           found /home/vadorovsky/go/bin/ginkgo, version 1.14.1
ok        golangci-lint    found /home/vadorovsky/go/bin/golangci-lint, version 1.31.0
ok        docker           found /usr/bin/docker
warning   docker-compose   docker-compose not found in $PATH
ok        vagrant          found /usr/bin/vagrant, version 2.2.10
warning   virtualbox       virtualbox not found in $PATH
ok        pip3             found /usr/bin/pip3, version 20.0.2
ok        docker-group     user vadorovsky in docker group

See https://docs.cilium.io/en/latest/contributing/development/dev_setup/
❯ which VirtualBox
/usr/bin/VirtualBox
❯ which virtualbox
which: no virtualbox in (/home/vadorovsky/repos/llvm-project/llvm/build/bin:/home/vadorovsky/opt/cross/bin:/home/vadorovsky/repos/llvm-project/llvm/build/bin:/home/vadorovsky/opt/cross/bin:/home/vadorovsky/bin:/usr/local/bin:/usr/bin:/bin:/home/vadorovsky/.cargo/bin:/home/vadorovsky/go/bin:/home/vadorovsky/go/src/k8s.io/kubernetes/third_party/etcd:/home/vadorovsky/.npm_packages/bin:/home/vadorovsky/.cargo/bin:/home/vadorovsky/go/bin:/home/vadorovsky/go/src/k8s.io/kubernetes/third_party/etcd:/home/vadorovsky/.npm_packages/bin)

**Review by vadorovsky — CHANGES_REQUESTED** [review]
Overall looks good, just one comment

File: tools/dev-doctor/rootcmd.go (hunk 16 above)
Comment (twpayne) [reply tools/dev-doctor/rootcmd.go]: This is now done, although with a special case for allowing the `virtualbox` binary to be called `VirtualBox`. It's easy to add more alternative names if needed.

Doing true case sensitivity is a little tricky because you'd either have to call `exec.LookPath` for all 512 possible combinations (e.g. `vIrTuAlBoX`, `ViRtUaLbOx`, etc.) or write code to read all directories in `$PATH` and check for binaries with matching names.

Comment (vadorovsky) [reply tools/dev-doctor/rootcmd.go]: You are right. Thanks for the fix!

**Comment by twpayne** [comment]
@pchaigno @mrostecki @aanm I think this PR is ready for final review now (you all have outstanding requested changes which I believe I have addressed).

**Review by vadorovsky — APPROVED** [review]
Thanks!

**Review by aanm — APPROVED** [review]
LGTM. One small note `docker-compose` is not required for development.

**Review by pchaigno — APPROVED** [review]
@twpayne Did you see [this comment](https://github.com/cilium/cilium/pull/13772#discussion_r512896892)? I think we should investigate this further even if the NFS check is optional. We don't want people to think they need something if they don't.

That can be a follow up though :slightly_smiling_face:

**Review by aanm — APPROVED** [review]

**Comment by twpayne** [comment]
test-me-please

**Comment by christarazi** [comment]
test-me-please
